Code/Framework/AzFramework/AzFramework/Spawnable/InMemorySpawnableAssetContainer.cpp

```diff
@@ -94,97 +56,88 @@ namespace AzToolsFramework::Prefab::PrefabConversionUtils
             AZ::Data::Asset<AZ::Data::AssetData> asset =
-                AZ::Data::AssetManager::Instance().FindOrCreateAsset(info.m_assetId, info.m_assetType, AZ::Data::AssetLoadBehavior::Default);
+                AZ::Data::AssetManager::Instance().FindOrCreateAsset(assetInfo.m_assetId, assetInfo.m_assetType, AZ::Data::AssetLoadBehavior::Default);
 
             // Update the asset registered in the AssetManager with the data of our product from the Prefab Processor
             AZ::Data::AssetManager::Instance().AssignAssetData(spawnableAssetData.m_assets.back());
         }
 
         if (targetSpawnableIndex == NoTargetSpawnable)
         {
-            return AZ::Failure(AZStd::string::format("Failed to produce the target spawnable '%.*s'.", AZ_STRING_ARG(spawnableName)));
+            return AZ::Failure(AZStd::string::format("Failed to produce the target spawnable '%.*s'.", AZ_STRING_ARG(rootSpawnableName)));
         }
 
         if (loadReferencedAssets)
         {
             LoadReferencedAssets(spawnableAssetData);
         }
 
-        auto& spawnableAssetDataAdded = m_spawnableAssets.emplace(spawnableName, spawnableAssetData).first->second;
+        // Delay resolving aliases to guarantee that the depended spawnables are already registered.
+        for (auto spawnablePair : spawnables)
+        {
+            Spawnable* spawnable = spawnablePair.first;
+            const AZStd::string& spawnableName = spawnablePair.second;
+            SpawnableAssetUtils::ResolveEntityAliases(spawnable, spawnableName);
+        }
+
+        auto& spawnableAssetDataAdded = m_spawnableAssets.emplace(rootSpawnableName, spawnableAssetData).first->second;
         spawnableAssetDataAdded.m_spawnableAssetId = spawnableAssetDataAdded.m_assets[targetSpawnableIndex].GetId();
         return AZ::Success(spawnableAssetDataAdded.m_assets[targetSpawnableIndex]);
     }
 
     InMemorySpawnableAssetContainer::CreateSpawnableResult InMemorySpawnableAssetContainer::CreateInMemorySpawnableAsset(
-        AZStd::string_view prefabFilePath, AZStd::string_view spawnableName, bool loadReferencedAssets)
+        Spawnable* spawnable,
+        const AZ::Data::AssetId& assetId,
+        bool loadReferencedAssets,
+        const AZStd::string& rootSpawnableName)
     {
-        AZ::IO::Path relativePath = m_loaderInterface->GenerateRelativePath(prefabFilePath);
-        auto templateId = m_prefabSystemComponentInterface->GetTemplateIdFromFilePath(relativePath);
-        if (templateId == InvalidTemplateId)
-        {
-            return AZ::Failure(AZStd::string::format("Template with source path '%.*s' is not found.", AZ_STRING_ARG(prefabFilePath)));
-        }
-
-        return CreateInMemorySpawnableAsset(templateId, spawnableName, loadReferencedAssets);
+        AZ::Data::AssetInfo assetInfo;
+        assetInfo.m_assetId = assetId;
+        assetInfo.m_assetType = spawnable->GetType();
+        assetInfo.m_relativePath = rootSpawnableName;
```

Comment:
If this should be a path and not any string, using AZ::IO::Path and AZ::IO::Pathview would be a better idea to avoid accepting incorrect paths.

Reply:
Good catch! So unlike the name implies, m_relativePath aren't always paths on disk. For example, the in-memory level spawnable won't have a path on disk and just uses "Root.spawnable" as m_relativePath